File: archive/go-v3.1.0/internal/fileops/sync.go
package fileops

import (
	"path"
	"path/filepath"
	"strings"
)

// ChangeMarker prefixes every itemized line rsync emits. It exists because the
// same stdout also carries warnings, a remote motd, and (in openrsync) the
// deletion notices that ignore --out-format: a parser has to tell those apart
// from plan lines, not guess from shape.
const ChangeMarker = "RHOSTSYNC|"

// RejectSyncTarget decides whether a directory-sync destination is safe enough to
// write to — and, when --delete is set, safe enough to *prune* (§28: "sync is the
// operation most likely to cause accidental data loss").
//
// The rules are deliberately few and mechanical:
//
//  1. a destination is never empty, the root, a glob, or carrying control
//     characters (shared with single-file transfers);
//  2. with --delete, an absolute destination must have at least two path
//     components. `/srv` may be exactly where a database lives; `/srv/app/deploy`
//     is a directory some name was chosen for. Refusing the one-component case is
//     the difference between a mistake and a disaster — and it is explainable:
//     rhost says so, and syncing into a subdirectory is the way round it;
//  3. with --delete, `~` or `~user` alone (someone's entire home) is refused for
//     the same reason;
//  4. `.` and `..` are refused always: what they name depends on the remote
//     tool's working directory, which rhost does not control.
func RejectSyncTarget(destination string, delete bool) error {
	if err := ValidateRemotePath(destination); err != nil {
		return err
	}
	destination = path.Clean(destination)
	if err := checkDestination(destination); err != nil {
		return err
	}
	switch destination {
	case ".", "./", "..", "../":
		return reject("sync destination %q depends on the remote working directory; use a path under the home directory or an absolute path", destination)
	}
	if !delete {
		return nil
	}
	if destination == "~" || destination == "~/" {
		return reject("--delete refuses an entire home directory (%q); choose a subdirectory", destination)
	}
	if strings.HasPrefix(destination, "~") && !strings.Contains(strings.TrimPrefix(destination, "~"), "/") {
		return reject("--delete refuses an entire home directory (%q); choose a subdirectory", destination)
	}
	if filepath.IsAbs(destination) && len(strings.Split(strings.Trim(destination, "/"), "/")) == 1 {
		return reject("--delete refuses a top-level directory (%q); sync into a subdirectory instead", destination)
	}
	return nil
}

// SyncOptions describes one directory sync.
//
// The source's *contents* are transferred: the trailing slash is added here
// rather than left to the caller, because "src" and "src/" mean different things
// to rsync and rhost must not depend on which one a human typed.
type SyncOptions struct {
	Source      string
	Destination string
	Delete      bool
	DryRun      bool
	Excludes    []string
}

// SyncArgs builds the rsync argv for one local→remote sync. sshOpts is the same
// `-o Key=Value` list rhost passes to ssh, so the transfer rides the existing
// ControlMaster instead of authenticating again (§8).
//
// rsync splits its own -e string on whitespace, so an option containing a space
// cannot travel that way: rhost then syncs over a fresh connection rather than
// mangling the argument, and reports which happened through multiplexed.
func SyncArgs(host string, o SyncOptions, sshOpts []string) (args []string, remote string, multiplexed bool, err error) {
	if err := RejectSyncTarget(o.Destination, o.Delete); err != nil {
		return nil, "", false, err
	}
	src, err := LocalArg(o.Source)
	if err != nil {
		return nil, "", false, err
	}

	args = []string{"--recursive"}
	// --times preserves mtimes, so a later sync is incremental rather than "the
	// whole tree changed". --no-owner/--no-group are explicit rather than left to
	// a default: preserving ownership would need root on the remote, and a sync
	// must not quietly fail at it. Symlinks are skipped rather than followed,
	// which is what both GNU rsync and the openrsync shipped by macOS do without
	// --links (§28: do not follow unexpected symlinks across boundaries).
	args = append(args, "--times", "--no-owner", "--no-group", "--no-perms")
	if o.DryRun {
		args = append(args, "--dry-run")
	}
	if o.Delete {
		args = append(args, "--delete")
	}
	for _, pat := range o.Excludes {
		if strings.ContainsAny(pat, "\n\x00") {
			return nil, "", false, reject("exclude pattern %q contains a newline or NUL", pat)
		}
		args = append(args, "--exclude", pat)
	}
	args = append(args, "--itemize-changes", "--out-format="+ChangeMarker+"%i|%n")

	shell, usable := remoteShell(sshOpts)
	args = append(args, "-e", shell)

	srcArg := strings.TrimRight(src, "/") + "/"
	args = append(args, srcArg, RemoteSpec(host, o.Destination))
	return args, RemoteSpec(host, o.Destination), usable, nil
}

// remoteShell renders rsync's -e value and reports whether it still carries rhost's
// options (false means the transfer cannot reuse the multiplexed connection).
//
// rsync splits its own -e string with shell-like rules, so an option containing
// whitespace is single-quoted instead of dropped: without the quotes the option
// would reach ssh as two arguments, and the old fallback turned that into a whole
// non-multiplexed transfer. Quoting was verified against GNU rsync and against
// the openrsync macOS ships; the shell-style `'\”` escape was *not* (openrsync
// rejects it as an unterminated quote), so an option containing a single quote
// still falls back rather than being mangled into something nearer to a bug.
func remoteShell(sshOpts []string) (string, bool) {
	quoted := make([]string, 0, len(sshOpts)+1)
	for _, o := range sshOpts {
		if strings.ContainsAny(o, "\n\x00") {
			return "ssh", false
		}
		q, ok := quoteRSH(o)
		if !ok {
			return "ssh", false
		}
		quoted = append(quoted, q)
	}
	if len(quoted) == 0 {
		return "ssh", false
	}
	return "ssh " + strings.Join(quoted, " "), true
}

// quoteRSH single-quotes an option that contains whitespace, which rsync's own
// splitter removes again before ssh sees the argument. An option that cannot be
// represented that way reports false, and the caller syncs over a fresh
// connection instead.
func quoteRSH(o string) (string, bool) {
	if !strings.ContainsAny(o, " \t") {
		return o, true
	}
	if strings.Contains(o, "'") {
		return "", false
	}
	return "'" + o + "'", true
}

// Action is what a change line says is about to happen. Coarse on purpose: the
// itemize string is kept beside it so nothing is hidden, while an agent can still
// branch on a small stable set.
type Action string

const (
	ActionCreate    Action = "create"
	ActionUpdate    Action = "update"
	ActionDelete    Action = "delete"
	ActionDirectory Action = "directory"
	ActionSkip      Action = "skip"
	ActionOther     Action = "other"
)

// Change is one line of rsync's itemized plan.
type Change struct {
	Action  Action `json:"action"`
	Path    string `json:"path"`
	Itemize string `json:"itemize"`
}

// ParseChanges turns rsync's stdout into structured changes. Lines without the
// marker (warnings, openrsync's delete notices) are returned separately so a
// human can still see what the tool said; they are never dropped.
func ParseChanges(out string) (changes []Change, other []string) {
	for _, line := range strings.Split(strings.ReplaceAll(out, "\r\n", "\n"), "\n") {
		switch {
		case line == "":
			continue
		case strings.HasPrefix(line, ChangeMarker):
			// The itemize column is fixed-width text with no "|" in it, so the
			// first separator is the only one to split on: everything after it is
			// the filename, "|" characters and all.
			itemize, name, found := strings.Cut(strings.TrimPrefix(line, ChangeMarker), "|")
			if !found {
				other = append(other, line)
				continue
			}
			changes = append(changes, Change{Action: classify(itemize), Path: name, Itemize: itemize})
		case strings.HasPrefix(line, "*deleting"):
			// openrsync prints deletions in its own format; GNU rsync puts them
			// through --out-format, so this branch is the other half of one rule.
			name := strings.TrimSpace(strings.TrimPrefix(line, "*deleting"))
			changes = append(changes, Change{Action: ActionDelete, Path: name, Itemize: "*deleting"})
		default:
			other = append(other, line)
		}
	}
	return changes, other
}

// classify maps an rsync itemize string to an Action.
//
// Only the leading characters are load-bearing: `>f` is a file being sent to the
// destination, `cd`/`.d` a directory, and a run of `+` in the transfer column
// means "created" while letters mean "something about it changed".
func classify(itemize string) Action {
	switch {
	case strings.HasPrefix(itemize, "*deleting"):
		return ActionDelete
	// The second character names the file type, and that is what decides whether
	// the entry is a file at all.
	case len(itemize) > 1 && itemize[1] == 'd':
		return ActionDirectory
	case len(itemize) > 1 && strings.IndexByte("Lcsbp-", itemize[1]) >= 0:
		// A symlink, socket, device or special file is neither copied nor
		// followed (§28: nothing follows a symlink across the boundary silently),
		// so it is reported as skipped rather than looking like a missing file.
		return ActionSkip
	case len(itemize) > 2 && strings.HasPrefix(itemize, ">f"):
		if strings.Trim(itemize[2:], "+") == "" {
			return ActionCreate
		}
		return ActionUpdate
	case len(itemize) > 2 && strings.HasPrefix(itemize, "cf"):
		return ActionCreate
	default:
		return ActionOther
	}
}

File: src/fileops/sync.rs
//! Directory sync: which destination `--delete` may prune, and the `rsync` argv.

use super::args::{check_destination, local_arg, remote_spec, validate_remote_path};
use super::changes::CHANGE_MARKER;
use super::{Refusal, reject};
use std::path::Path;

/// Decides whether a directory-sync destination is safe to write to — and, with
/// `--delete`, safe to prune. The rules are few and mechanical.
pub fn reject_sync_target(destination: &str, delete: bool) -> Result<(), Refusal> {
    validate_remote_path(destination)?;
    let cleaned = clean_remote_path(destination);
    check_destination(&cleaned)?;
    if matches!(cleaned.as_str(), "." | "./" | ".." | "../") {
        return Err(reject(format!(
            "sync destination {cleaned:?} depends on the remote working directory; use a path under the home directory or an absolute path"
        )));
    }
    if !delete {
        return Ok(());
    }
    if cleaned == "~" || cleaned == "~/" {
        return Err(reject(format!(
            "--delete refuses an entire home directory ({cleaned:?}); choose a subdirectory"
        )));
    }
    if let Some(rest) = cleaned.strip_prefix('~') {
        if !rest.contains('/') {
            return Err(reject(format!(
                "--delete refuses an entire home directory ({cleaned:?}); choose a subdirectory"
            )));
        }
    }
    if Path::new(&cleaned).is_absolute() && cleaned.trim_matches('/').split('/').count() == 1 {
        return Err(reject(format!(
            "--delete refuses a top-level directory ({cleaned:?}); sync into a subdirectory instead"
        )));
    }
    Ok(())
}

/// Lexically cleans a remote path the way the remote shell's tools treat it:
/// repeated separators collapse, `.` disappears, and `..` cancels the component
/// before it. A remote path is POSIX on every host, so this is deliberately not
/// the platform's own path rules (a host is never a Windows path).
///
/// The cleaning happens before the `--delete` rules are judged, because
/// `"/tmp/../"` names the filesystem root however it was typed.
fn clean_remote_path(path: &str) -> String {
    if path.is_empty() {
        return ".".to_string();
    }
    let rooted = path.starts_with('/');
    let mut parts: Vec<&str> = Vec::new();
    for part in path.split('/') {
        match part {
            "" | "." => {}
            ".." => match parts.last() {
                Some(last) if *last != ".." => {
                    parts.pop();
                }
                _ if !rooted => parts.push(".."),
                _ => {}
            },
            other => parts.push(other),
        }
    }
    let joined = parts.join("/");
    if rooted {
        format!("/{joined}")
    } else if joined.is_empty() {
        ".".to_string()
    } else {
        joined
    }
}

/// One directory sync. The direction decides which endpoint the destination
/// guard judges, because that is the side `--delete` can prune.
pub struct SyncRequest {
    pub host: String,
    pub source: String,
    pub destination: String,
    pub delete: bool,
    pub dry_run: bool,
    pub excludes: Vec<String>,
}

/// The rsync argv for one local→remote sync, the remote spec it names, and
/// whether the transfer still carries rhost's own options.
pub struct SyncPlan {
    pub args: Vec<String>,
    /// The endpoint the plan reports, with a trailing slash: the *contents* of a
    /// source are transferred, and the reported value says so.
    pub source: String,
    pub destination: String,
    pub multiplexed: bool,
}

fn transfer_options(request: &SyncRequest) -> Result<Vec<String>, Refusal> {
    let mut args = vec![
        "--recursive".to_string(),
        // --times preserves mtimes so a later sync is incremental rather than
        // "the whole tree changed"; ownership and permissions are explicit
        // rather than left to a default that would need root remotely.
        "--times".to_string(),
        "--no-owner".to_string(),
        "--no-group".to_string(),
        "--no-perms".to_string(),
    ];
    if request.dry_run {
        args.push("--dry-run".to_string());
    }
    if request.delete {
        args.push("--delete".to_string());
    }
    for pattern in &request.excludes {
        if pattern.contains('\n') || pattern.contains('\0') {
            return Err(reject(format!(
                "exclude pattern {pattern:?} contains a newline or NUL"
            )));
        }
        args.push("--exclude".to_string());
        args.push(pattern.clone());
    }
    args.push("--itemize-changes".to_string());
    args.push(format!("--out-format={CHANGE_MARKER}%i|%n"));
    Ok(args)
}

/// Local directory → remote directory.
pub fn push_args(request: &SyncRequest, ssh_options: &[String]) -> Result<SyncPlan, Refusal> {
    reject_sync_target(&request.destination, request.delete)?;
    let source = local_arg(&request.source)?;
    let mut args = transfer_options(request)?;
    let (shell, multiplexed) = remote_shell(ssh_options);
    args.push("-e".to_string());
    args.push(shell);
    // The source's *contents* are transferred; the trailing slash is added here
    // because "src" and "src/" mean different things to rsync.
    let source = format!("{}/", source.trim_end_matches('/'));
    args.push(source.clone());
    let remote = remote_spec(&request.host, &request.destination);
    args.push(remote.clone());
    Ok(SyncPlan {
        args,
        source,
        destination: remote,
        multiplexed,
    })
}

/// Remote directory → local directory. The risk is reversed here: `--delete`
/// threatens *local* files, so the destination guard judges the local path.
pub fn pull_args(request: &SyncRequest, ssh_options: &[String]) -> Result<SyncPlan, Refusal> {
    reject_sync_target(&request.destination, request.delete)?;
    let destination = local_arg(&request.destination)?;
    let mut args = transfer_options(request)?;
    let (shell, multiplexed) = remote_shell(ssh_options);
    args.push("-e".to_string());
    args.push(shell);
    let source = remote_spec(
        &request.host,
        &format!("{}/", request.source.trim_end_matches('/')),
    );
    args.push(source.clone());
    args.push(destination.clone());
    Ok(SyncPlan {
        args,
        source,
        destination,
        multiplexed,
    })
}

/// Renders rsync's `-e` value and reports whether it still carries rhost's
/// options. rsync splits its own `-e` string with shell-like rules, so an option
/// containing whitespace is single-quoted instead of dropped; an option that
/// cannot be represented that way reports `false`, and the transfer runs over a
/// fresh connection rather than mangled argv.
fn remote_shell(ssh_options: &[String]) -> (String, bool) {
    let mut quoted: Vec<String> = Vec::with_capacity(ssh_options.len() + 1);
    for option in ssh_options {
        if option.contains('\n') || option.contains('\0') {
            return ("ssh".to_string(), false);
        }
        match quote_rsh(option) {
            Some(value) => quoted.push(value),
            None => return ("ssh".to_string(), false),
        }
    }
    if quoted.is_empty() {
        return ("ssh".to_string(), false);
    }
    (format!("ssh {}", quoted.join(" ")), true)
}

fn quote_rsh(option: &str) -> Option<String> {
    if !option.contains([' ', '\t']) {
        return Some(option.to_string());
    }
    if option.contains('\'') {
        return None;
    }
    Some(format!("'{option}'"))
}
